synapse/lib/cell.py

```diff
@@ -2619,6 +2646,80 @@ def _loadCellYaml(self, *path):
 
         return {}
 
+    def _hasEasyPerm(self, item, user, level):
+
+        if user.isAdmin():
+            return True
+
+        userlevel = item['permissions']['users'].get(user.iden)
+        if userlevel is not None:
+            if userlevel == 0:
+                return False
+            elif userlevel >= level:
+                return True
+
+        roleperms = item['permissions']['roles']
+        for role in user.getRoles():
+            rolelevel = roleperms.get(role.iden)
+            if rolelevel is not None:
+                if rolelevel == 0:
+                    return False
+                elif rolelevel >= level:
+                    return True
+
+        # allow read by default unless explicitly set
+        if level <= PERM_READ:
+            return True
+
+        return False
+
+    def _reqEasyPerm(self, item, user, level, mesg=None):
+        '''
+        Require the user (or an assigned role) to have the given permission
+        level on the specified item. The item must implement the "easy perm"
+        convention by having a key named "permissions" which adheres to the
+        easyPermSchema definition.
+
+        NOTE: By default a user will only be denied read access if they
+              (or an assigned role) has PERM_DENY assigned.
+        '''
+        if self._hasEasyPerm(item, user, level):
+            return
+
+        if mesg is None:
+            permname = permnames.get(level)
+            mesg = f'User has insufficient permissions (requires: {permname}).'
+
+        raise s_exc.AuthDeny(mesg=mesg)
+
+    async def _setEasyPerm(self, item, scope, iden, level):
+        '''
+        Set a user or role permission level within an object that uses the "easy perm"
+        convention. If level is None, permissions are removed.
+        '''
+        if scope not in ('users', 'roles'):
+            raise s_exc.BadArg(mesg=f'Invalid permissions scope: {scope}')
+
+        if scope == 'users':
+            await self.auth.reqUser(iden)
+
+        elif scope == 'roles':
+            await self.auth.reqRole(iden)
+
+        perms = item['permissions'].get(scope)
+        if level is None:
+            perms.pop(iden, None)
+        else:
+            perms[iden] = level
```

Comment:
a level that is not in the union of `None` and the known four permissions will be set.
So we can set `-1` and `5` as levels. That doesn't seem correct.

Reply:
Yeah this was getting checked by the jsonschema but probably better to check here

Reply:
It feels like any Cell level API that tags level should validate it, much like we validate scope. That would mean we'd be able to avoid checking it in multiple other places too.